spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java

```diff
@@ -0,0 +1,215 @@
+		return this.pulsarConsumerFactory;
+	}
+
+	public ReactivePulsarContainerProperties getContainerProperties() {
+		return this.pulsarContainerProperties;
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	protected void setRunning(boolean running) {
+		this.running = running;
+	}
+
+	@Override
+	public void setupMessageHandler(Object messageHandler) {
+		this.pulsarContainerProperties.setMessageHandler(messageHandler);
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	@Override
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	@Override
+	public final void start() {
+		synchronized (this.lifecycleMonitor) {
+			if (!isRunning()) {
+				Assert.state(
+						this.pulsarContainerProperties.getMessageHandler() instanceof ReactivePulsarStreamingHandler<?>
+								|| this.pulsarContainerProperties
+										.getMessageHandler() instanceof ReactivePulsarMessageHandler<?>,
+						() -> "A ReactivePulsarMessageHandler or ReactivePulsarStreamingHandler implementation must be provided");
+				doStart();
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		synchronized (this.lifecycleMonitor) {
+			if (isRunning()) {
+				doStop();
+			}
+		}
+	}
+
+	public ReactiveMessageConsumerBuilderCustomizer<T> getConsumerCustomizer() {
+		return this.consumerCustomizer;
+	}
+
+	@Override
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	public void setConsumerCustomizer(ReactiveMessageConsumerBuilderCustomizer consumerCustomizer) {
+		this.consumerCustomizer = consumerCustomizer;
+	}
+
+	private void doStart() {
+		this.pipeline = startPipeline(this.pulsarContainerProperties);
+		setRunning(true);
+	}
+
+	public void doStop() {
+		try {
+			this.logger.info("Closing this reactive pipeline.");
+			this.pipeline.close();
+		}
+		catch (Exception e) {
+			this.logger.error(e, () -> "Error closing Pulsar Reactive pipeline.");
+		}
+	}
+
+	@Override
+	public void destroy() {
+
+	}
+
+	@SuppressWarnings({ "unchecked" })
+	private ReactiveMessagePipeline startPipeline(ReactivePulsarContainerProperties containerProperties) {
+		ReactiveMessageConsumerBuilderCustomizer<T> customizer = (builder) -> {
+			if (containerProperties.getSubscriptionType() != null) {
+				builder.subscriptionType(containerProperties.getSubscriptionType());
+			}
+			if (containerProperties.getSubscriptionName() != null) {
+				builder.subscriptionName(containerProperties.getSubscriptionName());
+			}
+			if (containerProperties.getTopics() != null) {
+				List<String> topicNames = Arrays.asList(containerProperties.getTopics());
+				if (!topicNames.isEmpty()) {
+					builder.topicNames(topicNames);
+				}
+			}
+			if (containerProperties.getTopicsPattern() != null) {
+				builder.topicsPattern(Pattern.compile(containerProperties.getTopicsPattern()));
```

Comment:
This makes me think we should/could do this "to pattern" earlier in the container props - if its invalid it would barf earlier.

Reply:
done